### PR TITLE
Update Optimus Adapter workflow 

### PR DESCRIPTION
**Pull request by rexwangcc**
### Purpose
_Please explain the purpose of this PR and include links to any GitHub issues that it fixes:_

- https://app.zenhub.com/workspaces/dcp-backlogs-5ac7bcf9465cb172b77760d9/issues/humancellatlas/secondary-analysis/368
---
### Changes
_Please list out what major changes were made in this PR to address the issue:_

- Update the AdapterOptimus WDL so it expects Zarr outputs from the Optimus analysis pipeline.
- Update the options of AdapterOptimus WDL.
- Update the `get_workflow_metadata.py` so it bypasses the error in CaaS when the workflow's content of metadata is too large.

---
### Review Instructions
_Please provide instructions about how should a reviewer test/verify the changes in this PR:_

#### Happy Path 1: 4k_pbmc with i1

~~- Please go to https://job-manager.caas-prod.broadinstitute.org/jobs/9fad12e1-ae75-4b41-bc6f-94d2b707f191 to check the workflow succeeded, which was kicked off for testing this adapter workflow.~~
~~- Please go to https://integration.data.humancellatlas.org/explore/files?filter=%5B%7B%22facetName%22%3A%22project%22%2C%22terms%22%3A%5B%22Optimus+Pipeline+Testing+Dataset+4kpbmc%22%5D%7D%5D and verify that the results of this run are shown properly.~~
~~- Please note the inputs to the above testing workflow is the full 4k_pbmc dataset.~~

- Please go to https://job-manager.caas-prod.broadinstitute.org/jobs/48506c11-7851-4391-b589-e1518646500b to check the workflow succeeded, which was kicked off for testing this adapter workflow.
- Please note the inputs to the above testing workflow is the full 4k_pbmc dataset.

#### Unhappy Path: 4k_pbmc without i1
~~- Please take a look at https://job-manager.caas-prod.broadinstitute.org/jobs/8be84893-6871-494b-8bc8-aa0c56b9fb40~~
~~- It ran through the `prep` task, which means the input preparation is OK. It failed at the analysis step.~~

- Please take a look at https://job-manager.caas-prod.broadinstitute.org/jobs/fb323315-ce47-4ac3-b482-f7692b96f76a to check the workflow succeeded, which was kicked off for testing this adapter workflow.
- It ran through the `prep` task, which means the input preparation is OK. It also invokes the "AttachBarcodesNoIndex" in Optimus analysis workflow, which indicates the `i1_fastq` was not passed in to the analysis pipeline as expected.

#### happy Path 2: 8k_pbmc with i1

Synced up with @kishorikonwar and decided to not un 8k dataset for this PR since it's captured by the test in Skylab already!

---
### PR Checklist
_Please ensure the following when opening a PR:_

- [ ] This PR added or updated tests.
- [ ] This PR updated docstrings or documentation.
- [ ] This PR applied code style guidelines:
    - [Google style docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google) Python code.
    - Mint WDL style guide for WDL.
- [x] This PR considered generalizability beyond our own use case.

---
### Follow-up Discussions
_Please append follow-up discussions and issues during the review process below:_

@pshapiro4broad suggested a better way to mimic `None` in WDL draft - 2, the latest change should reflect that. correspsonding tests:
- No index: https://job-manager.caas-prod.broadinstitute.org/jobs/dbd0aa72-0d37-44a1-906c-a76c37059423 (It ran thru the testing step but failed at a tep that talking to the external service, due to a PAPI v2 error, which shouldn't block us from merging!)
- With index: https://job-manager.caas-prod.broadinstitute.org/jobs/5689c99a-5bf4-417e-a99c-97796199ab4c

### CODE REVIEW

**Comment by samanehsan** [comment]
As we just discussed offline, I'd like to see a test of this adapter workflow with data that does not have an I1 file to make sure that case passes! 🙂

**Comment by rexwangcc** [comment]
@samanehsan Could you take another look at this?

File: pipeline_tools/input_utils.py [hunk 1]
```diff
@@ -236,10 +236,6 @@ def create_optimus_input_tsv(uuid, version, dss_url):
     with open('i1.txt', 'w') as f:
         for url in i1_urls:
             f.write(url + '\n')
-    with open('lanes.txt', 'w') as f:
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply pipeline_tools/input_utils.py]: Good question, so we have never used this information in Optimus actually. This was a required info for cellranger, it's just copied over. check: https://broadinstitute.atlassian.net/jira/software/projects/GH/boards/530?selectedIssue=GH-56

File: adapter_pipelines/cellranger/adapter.wdl [hunk 2]
```diff
@@ -150,7 +150,7 @@ workflow Adapter10xCount {
   Int max_cromwell_retries = 0
   Boolean add_md5s = false
 
-  String pipeline_tools_version = "v0.46.2"
+  String pipeline_tools_version = "rex-adapter-optimus"
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply adapter_pipelines/cellranger/adapter.wdl]: Yes! I'm going to use `v0.46.3`before merging and then tag the repo with `v0.46.3`

File: adapter_pipelines/Optimus/adapter.wdl [hunk 3]
```diff
@@ -38,16 +38,16 @@ task GetInputs {
...
     Array[File] http_responses = glob("response_*.txt")
   }
 }
 
-task inputs_for_submit {
-    Array[File] r1_fastq
-    Array[File] r2_fastq
-    Array[File] i1_fastq
+task InputsForSubmit {
+    Array[String] r1_fastq
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply adapter_pipelines/Optimus/adapter.wdl]: Thanks for this, I searched for it but failed to find it! 

In this specific case, I think we might be fine with using `String`. The file name of the fastq files in this step will always be in the format of `gs://org-humancellatlas-dss-checkout-integration/bundles/a4fbfc4f-5309-473b-baf9-7bb02f4c9d9e.2019-02-27T025644.950428Z/xxxxxxx.fastq.gz`, which should be sufficient for the call caching checkers.

[the remark this answers is not in the document]
Comment (rexwangcc) [reply adapter_pipelines/Optimus/adapter.wdl]: I take the previous comment back. @ApChagi and @samanehsan identified passing a `Array[File]?` output as an `Array[String]?` input to another task, like I'm doing here, caused some weird behaivors to the `read_lines()` WDL function, I will change it and use the `localization_optional: true`!!!

File: adapter_pipelines/Optimus/adapter.wdl [hunk 4]
```diff
@@ -101,6 +106,11 @@ workflow AdapterOptimus {
   File annotations_gtf  # gtf containing annotations for gene tagging
   File ref_genome_fasta  # genome fasta file
   String fastq_suffix = ".gz"  # add this suffix to fastq files for picard
+  
+  # Note: This "Empty" is a workaround in WDL-draft to simulate a "None" type
+  # once the official "None" type is introduced (probably in WDL 2.0)
+  # we should consider replace this workaround
+  Array[File]? Empty
```
Comment (pshapiro4broad) [inline adapter_pipelines/Optimus/adapter.wdl]: Maybe call it `None`? Will make migrating to the real `None` even easier, when it arrives.
```suggestion
  Array[File]? None
```

Comment (rexwangcc) [reply adapter_pipelines/Optimus/adapter.wdl]: Good point, will do!

**Pull request merged**